sunkit_image/nafe.py

```diff
@@ -0,0 +1,411 @@
+           [0.65289406, 0.73715583, 0.7741782 , 0.73715583, 0.65289406]])
+    """
+    M = len(wl)
+    sigma = [2 ** (m / 2) for m in range(1, M+1)]  # starts from 1 for maths
+    d = 2 * np.pi / np.sum(wl * np.power(sigma, -1/2))
+
+    wlG = 0
+    for s, w in zip(sigma, wl):
+        G = lambda r, c: np.exp(-0.5 * ((r - n//2) ** 2 + (c - n//2) ** 2) * (s ** -2)) / (2 * np.pi * np.sqrt(s))
+        wlG += w * np.fromfunction(G, (n, n))
+
+    return d * wlG
+
+
+def _transform(old_margin, new_margin, old_value, power=1.0):
+    """
+    # TODO: add doc
+    """
+    return new_margin[0] + (new_margin[1] - new_margin[0]) * (((old_value - old_margin[0]) / (old_margin[1] - old_margin[0])) ** (1.0 / power))
+
+
+class _Nafe():
+    """
+    Utility class for caching shared variables for multiprocessing of nafe.
+    """
+    def __init__(self, image, nbins, gamma, sigma, weight, imargin, n, c):
+        self.image = image
+        self.nbins = nbins
+        self.gamma = gamma
+        self.sigma = sigma
+        self.weight = weight
+        self.imargin = imargin
+        self.omargin = (0, 2 ** 16 - 1)
+
+        self.halfn = n // 2
+        self.membership = _get_membership_grade(n=n, wl=c)
+
+    def transform(self, index):
+        row, col = index
+        rows, cols = self.image.shape
+
+        img_minrow = max(0, row - self.halfn)
+        img_mincol = max(0, col - self.halfn)
+        img_maxrow = min(rows, row + self.halfn + 1)
+        img_maxcol = min(cols, col + self.halfn + 1)
+
+        mem_minrow = img_minrow - row + self.halfn
+        mem_mincol = img_mincol - col + self.halfn
+        mem_maxrow = img_maxrow - row + self.halfn
+        mem_maxcol = img_maxcol - col + self.halfn
+
+        neighborhood = self.image[img_minrow: img_maxrow, img_mincol: img_maxcol]
+        membership = self.membership[mem_minrow: mem_maxrow, mem_mincol: mem_maxcol]
+
+        hist, bins = np.histogram(neighborhood, bins=self.nbins, weights=membership)
+        bins = (bins[:-1] + bins[1:]) / 2
+
+        cumsum_hist = np.cumsum(hist)
+        cumsum_hist /= cumsum_hist[-1]  # normalize
+        # add artificial noise, reshaping to dummy 2d to enable 2d gauss
+        # from skimage then flattening again
+        cum_fuzzy_hist = skimage.filters.gaussian(cumsum_hist.reshape(self.nbins, 1), sigma=self.sigma).flatten()
+
+        old_val = self.image[row, col]
+        gamma_val = _transform(old_margin=self.imargin, new_margin=self.omargin, old_value=old_val, power=self.gamma)
+        nafe_val = np.interp(self.image[row, col], bins, cum_fuzzy_hist)  # returns [0., 1.]
+        nafe_val = _transform(old_margin=(0., 1.), new_margin=self.omargin, old_value=nafe_val)
+
+        return (1 - self.weight) * gamma_val + self.weight * nafe_val
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    # Required
+    parser.add_argument("-i", "--image", help="path to input image", required=True)
```

Comment:
normally, required arguments as the image in this case don't use a flag, but as a positional argument.

Reply:
Yes, missed this.